### DISCUSSION TITLE
Middleware runs twice when endpoint is called without trailing "/" (but defined with)

### DISCUSSION THREAD

**Discussion by jmwielandt** [fastapi/fastapi#8607] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import asyncio
import signal

from fastapi import APIRouter, FastAPI, Request
import hypercorn
from hypercorn.asyncio import serve
from starlette.middleware.base import BaseHTTPMiddleware


app = FastAPI()
router = APIRouter(prefix="/example")

STATE = 0  # dummy example


@router.get("/")
async def handle_example():
    print("I'm in!")
    return {"num_of_visits": STATE}


# middleware
async def log_request(request: Request, call_next):
    path = request.url.path
    method = request.method
    global STATE
    STATE += 1
    print(f"Logging request performed. Path={path!r}. Method={method!r}. State={STATE}")
    return await call_next(request)


def main():
    app.include_router(router)
    app.add_middleware(BaseHTTPMiddleware, dispatch=log_request)

    signal.signal(signal.SIGINT, signal.SIG_DFL)

    loop = asyncio.new_event_loop()

    cfg = hypercorn.Config()
    cfg.bind = "localhost:3000"

    server_coro = serve(app, cfg, shutdown_trigger=lambda: asyncio.Future())

    loop.create_task(server_coro)

    loop.run_forever()


if __name__ == "__main__":
    main()
```

### Description

1. Create a pipenv environment with `pipenv --python 3.10`
2. Install dependencies: `pipenv install fastapi`, `pipevn install hypercorn`
3. Run the server: `pipenv run py main.py`  (assuming you named the file "main.py")
4. Call on your browser to the endpoint `localhost:3000/example` and compare the logs on console with the endpoint `localhost:3000/example/`  (notice the trailing slash `/`).

By calling the endpoint without the trailing /, the app calls two times the middleware instead of one.

### Operating System

Windows

### Operating System Details

Windows 10 Home Single Language
Version: 21H2
Build: 19044.1566
Windows Feature Experience Pack 120.2212.4170.0

### FastAPI Version

0.74.1

### Python Version

Python 3.10.2

### Additional Context

Hypercorn version: 0.13.2

Might be related to PR https://github.com/tiangolo/fastapi/pull/415
I've uploaded all into [this repository](https://github.com/jmwielandt/mre-big-fastapi), so you can just clone it, install dependencies and run.

![imagen](https://user-images.githubusercontent.com/37153467/156408897-c6167f72-b613-4685-8a99-d8f87e401ce1.png)

**Comment by upeshphable** [comment, marked as answer]
It would because of redirection from "non-slash" path to "slash" path. I'm not sure where redirection happens, but that is the most probable explanation I can think of.

**Comment by odiseo0** [comment]
I also think this is because of the redirection from non-slash to slash path, I don't know if this is bad or should FastAPI considerer not run the Middleware when the response code is 307 (Redirection)

**Comment by jmwielandt** [comment]
@odiseo0 idk at which point the "redirection" occurs, but in my opinion is really bad. In my example is just a visits counter, but if yor middleware does anything more critical it could put your application in an unstable state (eg. lock.aquire) or impact performance (eg. double database querying)


Imagine a CRUD for a "settings" resource. All four endpoints (GET, POST, PATCH, DELETE) should use the same lock to ensure consistency, so you decide, naively, to have one middleware which will acquire the lock and then each endpoint will release it.

![imagen](https://user-images.githubusercontent.com/37153467/157081245-ea51c04d-4a70-47f2-a413-a7061e1ba253.png)

But one of your customers calls your endpoint without the sacred /, so you tried to acquire the lock twice, the first one (with the redirect) was never released and also didn't called the endpoint so... your app is blocked until you restart it :D

![imagen](https://user-images.githubusercontent.com/37153467/157082119-3d03ae68-ed85-4887-9120-1369cc10f2b4.png)

**Comment by camitz** [comment]
The thing is, the redirect is to http, regardless of the original request. This is not detected on localhost, but in production it's forbidden (mixed content). 
It took an hour for my team to find the missing slash.
So I'd say this is really really bad.
